src/viam_mlmodelservice_triton_impl.cpp

```diff
@@ -338,6 +340,87 @@ class Service : public vsdk::MLModelService, public vsdk::Stoppable, public vsdk
         return state_;
     }
 
+    static std::string initialize_directory_(std::string model_name, std::int64_t version, bool is_tf) {
+        const char* base_directory = std::getenv("VIAM_MODULE_DATA");
+
+        std::stringstream ss;
+        ss << base_directory << "/" << model_name;
+        if (is_tf) {
+            // TensorFlow models need one extra directory to work with
+            // Triton. So, the symlink to the directory containing the
+            // data will be named "model.savedmodel" inside the version
+            // directory, rather than being the version directory
+            // itself.
+            ss << "/" << version;
+        }
+        const std::string directory_name = ss.str();
+
+        std::error_code ec;
+        bool success = std::filesystem::create_directories(directory_name, ec);
+        // If the directory already exists, `success` will be false but the
+        // error code will be 0. Don't treat that as an error.
+        if (!success && ec.value() != 0) {
+            std::ostringstream buffer;
+            buffer << service_name << ": cannot create directory structure";
+            throw std::runtime_error(buffer.str());
+        }
+        return directory_name;
+    }
+
+    static void symlink_mlmodel_(const struct state_& state) {
+        const auto& attributes = state.configuration.attributes();
+
+        auto model_path = attributes->find("model_path");
+        if (model_path == attributes->end()) {
+            std::ostringstream buffer;
+            buffer << service_name
+                   << ": Required parameter `model_path` not found in configuration";
+            throw std::invalid_argument(buffer.str());
+        }
+
+        const std::string* model_path_string = model_path->second->get<std::string>();
+        if (!model_path_string || model_path_string->empty()) {
+            std::ostringstream buffer;
+            buffer << service_name
+                   << ": Required non-empty string parameter `model_path` is either not "
+                      "a string or is an empty string";
+            throw std::invalid_argument(buffer.str());
+        }
+
+        // If the model_version is -1, it means use the most recent version. We don't want to add a
+        // directory named `-1`, so just use a large number.
```

Comment:
Rephrase: it's not that we don't want to add a directory named `-1`, it's that we cannot do so because trying it has errors.